gt/main.c

```diff
@@ -81,39 +155,107 @@ gt_proc(void *arg)
 			continue;
 
 		for (i = 0; i < num_rx; i++) {
+			int ret;
+			uint8_t priority;
+			struct rte_mbuf *m = rx_bufs[i];
+			struct gt_packet_fields fields;
+			struct ggu_policy policy;
+			struct ipip_tunnel_info tunnel;
+			struct ether_hdr *new_eth;
+
 			/*
-			 * TODO Decapsulate the packets.
+			 * Decapsulate the packets.
 			 *
 			 * Only request packets and priority packets
 			 * with capabilities about to expire go through a
 			 * policy decision.
 			 *
 			 * Other packets will be fowarded directly.
 			 */
+			ret = decapsulate(m, &priority, &tunnel);
+			if (ret < 0) {
+				rte_pktmbuf_free(m);
+				continue;
+			}
+
```

Comment:
Notice that GT blocks set RSS to receive packets, but RSS doesn't filter packets whose destination IP address is not the Grantor server itself. So we must test if the destination is not the GT server. You can use `likely()` or `unlikely()` to this test here. A failure must be reported as an ALERT as well.